crates/sui-core/src/authority/authority_store.rs

```diff
@@ -1014,6 +1053,17 @@ impl AuthorityStore {
             iter::once((transaction_digest, transaction.serializable_ref())),
         )?;
 
+        // If there were any deleted shared objects as input to this transaction, update the marker table with our digest to be used as dependency of
+        // the next transaction that takes this object as an input
+        for (obj_id, _) in inner_temporary_store.deleted_shared_object_keys.clone() {
+            let object_key = (epoch_id, ObjectKey(obj_id, SHARED_OBJECT_MARKER_VERSION));
+
+            write_batch.insert_batch(
+                &self.perpetual_tables.object_per_epoch_marker_table,
+                iter::once((object_key, MarkerValue::SharedDeleted(*transaction_digest))),
+            )?;
```

Comment:
I'm pretty sure we have a race here. Consider the following certs:

C1: Deletes O
C2 & C3: Both read O

Once C1 has been executed, C2 and C3 can run in either order - this is because of the implementation of `multi_input_objects_available`, which declares deleted inputs as available.

The fact that their execution order is un-constrained combined with this write means that some validators could declare C2 to be a dependency of C3, while others could do the reverse, which will cause a fork. (Note: transaction manager will constrain the ordering *only if they C2 and C3 are scheduled prior to C1 executing*)

I wrote up a quick test case to demonstrate this: https://github.com/MystenLabs/sui/commit/4a91f7ef213cf1d7d3913cf1a99418e5d71caac7

Unfortunately, my test isn't able to demonstrate this bug, because the fullnode panics when attempting to execute C2 (or C3) from checkpoint, because there are missing shared input locks.

So we have two problems:
1. How to resolve the dependency issue: It's not clear to me what the solution is since we don't have any way to enforce that the txns are executed in the same order as they were sequenced. This does suggest to me that we should take another look at the plan where C2 and C3 would depend only on C1, not on each other.
2. The first issue is quite a subtle issue, but I'm concerned that we can't even reach the point of testing it due to a much more basic failure. I think this points towards this PR being significantly under-tested.

Reply:
We need smearing in every level:
- When a tx that uses deleted shared object is committed, we write the new marker with the lamport version in the marker table, and notify based on the lamport version
- When TxManager is ordering a tx that uses deleted shared object, it waits for the lamport version

Reply:
This will require commiting the originally assigned shared input versions in effects, which we don't currently do. (Otherwise execute from checkpoint will not know what version to wait for)

Reply:
As mentioned in the channel, we are _no longer going with the plan where C2 and C3 depend only on C1._ C3 depends on C2 and C2 depends on C1, forcing an execution order that matches consensus ordering. The object being marked available does not mean that we override consensus execution order and execute C3 before C2 if consensus ordered them the other way. 

> I wrote up a quick test case to demonstrate this: https://github.com/MystenLabs/sui/commit/4a91f7ef213cf1d7d3913cf1a99418e5d71caac7

> Unfortunately, my test isn't able to demonstrate this bug, because the fullnode panics when attempting to execute C2 (or C3) from checkpoint, because there are missing shared input locks.

This is the the same potential issue I mentioned before in the shared object deletion channel, and you mentioned we will not attempt to execute out of consensus order and not reach this panic. Given the above change of plan, is this execution out of consensus order something we expect will happen and should explicitly prevent in this PR?

Reply:
> This will require committing the originally assigned shared input versions in effects, which we don't currently do.
 
 Somehow I remember we discussed this somewhere, where we would commit it in `shared_objects` field of effects, and use the special deleted object digest to represent deleted input?